File: solutions/cpp/0084.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int ret = 0;
        stack<int> stack;
        stack.push(-1);

        for (int i = 0; i <= heights.size(); i++) {
            while (stack.top() != -1 && (i == heights.size() || heights[i] < heights[stack.top()])) {
                int h = heights[stack.top()];
                stack.pop();
                ret = max(ret, h * (i - stack.top() - 1));
            }
            stack.push(i);
        }

        return ret;
    }
};

File: solutions/python3/0084.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        ret = 0
        stack = [-1]

        for i in range(len(heights) + 1):
            while stack[-1] != -1 and (i == len(heights) or heights[i] < heights[stack[-1]]):
                h = heights[stack.pop()]
                ret = max(ret, h * (i - stack[-1] - 1))
            stack.append(i)

        return ret